packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx

```diff
@@ -66,81 +68,95 @@ export function PresentationInstanceFilterDialog(props: PresentationInstanceFilt
-            <Button className="presentation-instance-filter-dialog-close-button" onClick={onClose}>
-              {translate("instance-filter-builder.cancel")}
-            </Button>
-          </Dialog.ButtonBar>
-        </div>
+        {descriptor instanceof Descriptor ? (
+          <PresentationInstanceFilterDialogContent {...restProps} descriptor={descriptor} onClose={onClose} />
+        ) : (
+          <DelayedCenteredProgressRadial />
+        )}
       </Dialog.Main>
     </Dialog>
   );
 }
 
-interface DelayLoadedPresentationInstanceFilterBuilderProps extends Omit<PresentationInstanceFilterBuilderProps, "descriptor"> {
-  descriptorGetter: () => Promise<Descriptor>;
-}
-
-function DelayLoadedPresentationInstanceFilterBuilder(props: DelayLoadedPresentationInstanceFilterBuilderProps) {
-  const { descriptorGetter, ...restProps } = props;
-  const descriptor = useDelayLoadedDescriptor(descriptorGetter);
-  if (!descriptor) {
-    return <DelayedCenteredProgressRadial />;
-  }
-
-  return <PresentationInstanceFilterBuilder {...restProps} descriptor={descriptor} />;
-}
-
-function useDelayLoadedDescriptor(descriptorGetter: () => Promise<Descriptor>) {
-  const [descriptor, setDescriptor] = useState<Descriptor>();
+function useDelayLoadedDescriptor(descr: Descriptor | (() => Promise<Descriptor>)) {
+  const [descriptor, setDescriptor] = useState<Descriptor | (() => Promise<Descriptor>)>(() => descr);
+  const disposed = useRef<boolean>(false);
 
   useEffect(() => {
-    let disposed = false;
     void (async () => {
-      const newDescriptor = await descriptorGetter();
-      // istanbul ignore else
-      if (!disposed) {
-        setDescriptor(newDescriptor);
+      if (descriptor && !(descriptor instanceof Descriptor)) {
+        const newDescriptor = await descriptor();
+        // istanbul ignore else
+        if (!disposed.current) {
+          setDescriptor(newDescriptor);
+        }
       }
     })();
     return () => {
-      disposed = true;
+      disposed.current = true;
     };
-  }, [descriptorGetter]);
+  }, [descr, descriptor]);
 
   return descriptor;
 }
 
-function DelayedCenteredProgressRadial() {
-  const [show, setShow] = useState(false);
+interface PresedntationInstanceFilterDialogContentProps extends Omit<PresentationInstanceFilterDialogProps, "isOpen" | "title" | "descriptor"> {
+  descriptor: Descriptor;
+}
 
-  useEffect(() => {
-    const timeout = setTimeout(() => {
-      setShow(true);
-    }, 250);
-    return () => {
-      clearTimeout(timeout);
-    };
-  }, []);
+function PresentationInstanceFilterDialogContent(props: PresedntationInstanceFilterDialogContentProps) {
+  const { onApply, initialFilter, descriptor, imodel, ruleGroupDepthLimit, filterResultCountRenderer, onClose } = props;
+
+  const { rootGroup, actions, validate } = usePropertyFilterBuilder({
+    initialFilter: initialFilter ? convertPresentationFilterToPropertyFilter(descriptor, initialFilter.filter) : undefined,
```

Comment:
We should avoid calling `convertPresentationFilterToPropertyFilter` on each render.

Reply:
fixed

Reply:
`convertPresentationFilterToPropertyFilter` is still called on each render